src/hotspot/share/opto/superword.cpp

```diff
@@ -3615,3 +3029,279 @@ bool SuperWord::same_origin_idx(Node* a, Node* b) const {
+    } else if (vtn->isa_ReductionVector() != nullptr) {
+      set_req_with_scalar(p0,   vtn, vtn_dependencies, 1); // scalar init
+      set_req_with_vector(pack, vtn, vtn_dependencies, 2); // vector
+    } else {
+      assert(vtn->isa_ElementWiseVector() != nullptr, "all other vtnodes are handled above");
+      if (VectorNode::is_scalar_rotate(p0) &&
+          p0->in(2)->is_Con() &&
+          Matcher::supports_vector_constant_rotates(p0->in(2)->get_int())) {
+        set_req_with_vector(pack, vtn, vtn_dependencies, 1);
+        set_req_with_scalar(p0,   vtn, vtn_dependencies, 2); // constant rotation
+      } else if (VectorNode::is_roundopD(p0)) {
+        set_req_with_vector(pack, vtn, vtn_dependencies, 1);
+        set_req_with_scalar(p0,   vtn, vtn_dependencies, 2); // constant rounding mode
+      } else if (p0->is_CMove()) {
+        // Cmp + Bool + CMove -> VectorMaskCmp + VectorBlend.
+        set_all_req_with_vectors(pack, vtn, vtn_dependencies);
+        VTransformBoolVectorNode* vtn_mask_cmp = vtn->in(1)->isa_BoolVector();
+        if (vtn_mask_cmp->test()._is_negated) {
+          vtn->swap_req(2, 3); // swap if test was negated.
+        }
+      } else {
+        set_all_req_with_vectors(pack, vtn, vtn_dependencies);
+      }
+    }
+
+    for (uint k = 0; k < pack->size(); k++) {
+      add_dependencies_of_node_to_vtn(pack->at(k), vtn, vtn_dependencies);
+    }
+  }
+}
+
+void SuperWordVTransformBuilder::build_edges_for_scalar_vtnodes(VectorSet& vtn_dependencies) {
+  for (int i = 0; i < _vloop_analyzer.body().body().length(); i++) {
+    Node* n = _vloop_analyzer.body().body().at(i);
+    VTransformScalarNode* vtn = get_vtnode(n)->isa_Scalar();
+    if (vtn == nullptr) { continue; }
+    vtn_dependencies.clear(); // Add every dependency only once per vtn.
+
+    if (n->is_Load()) {
+      set_req_with_scalar(n, vtn, vtn_dependencies, MemNode::Address);
+    } else if (n->is_Store()) {
+      set_req_with_scalar(n, vtn, vtn_dependencies, MemNode::Address);
+      set_req_with_scalar(n, vtn, vtn_dependencies, MemNode::ValueIn);
+    } else if (n->is_CountedLoop()) {
+      continue; // Is "root", has no dependency.
+    } else if (n->is_Phi()) {
+      // CountedLoop Phi's: ignore backedge (and entry value).
+      assert(n->in(0) == _vloop.cl(), "only Phi's from the CountedLoop allowed");
+      set_req_with_scalar(n, vtn, vtn_dependencies, 0);
+      continue;
+    } else {
+      set_all_req_with_scalars(n, vtn, vtn_dependencies);
+    }
+
+    add_dependencies_of_node_to_vtn(n, vtn, vtn_dependencies);
+  }
+}
+
+// Create a vtnode for each pack. No in/out edges set yet.
+VTransformVectorNode* SuperWordVTransformBuilder::make_vtnode_for_pack(const Node_List* pack) const {
+  uint pack_size = pack->size();
+  Node* p0 = pack->at(0);
+  int opc = p0->Opcode();
+  VTransformVectorNode* vtn = nullptr;
+
+  if (p0->is_Load()) {
+    vtn = new (_graph.arena()) VTransformLoadVectorNode(_graph, pack_size);
```

Comment:
You use `_graph.arena()` quite often. Maybe you want to create a new `_arena` field for easier access.

Reply:
I'll see if I can do this directly with the `new` operator...

Reply:
I think having a `_arena` field in the `SuperWordVTransformBuilder` would be confusing and not a good idea. People might think that this is an arena that is used during the life-cycle of the builder. But this `_graph.arena()`, now `_vtransform.arena()` is used for the life-cycle of the `vtransform`, which outlives the builder. We could have a `_vtransform_arena` field, but this is not really better than `_vtransform.arena()` IMHO.

Reply:
I agree that "arena" could be confused for an arena to be used for all kinds of stuff during the life-cycle for the class while it is the vtransform-arena. `_vtransform.arena()` is perfectly fine. I would have preferred a field since you access it a lot but I guess it's just personal taste ;-)